run/testing/sigterm_handler.e2e_test.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudruntests
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/logging/logadmin"
+	"google.golang.org/api/iterator"
+
+	"github.com/GoogleCloudPlatform/golang-samples/internal/cloudrunci"
+	"github.com/GoogleCloudPlatform/golang-samples/internal/testutil"
+)
+
+func TestSigtermHandlerService(t *testing.T) {
+	tc := testutil.EndToEndTest(t)
+
+	runID := time.Now().Format("20060102-150405")
+	service := cloudrunci.NewService("sigterm-handler", tc.ProjectID)
+	service.Dir = "../sigterm-handler"
+	if err := service.Deploy(); err != nil {
+		t.Fatalf("service.Deploy %q: %v", service.Name, err)
+	}
+	defer GetLogEntries(service, runID, tc.ProjectID, t)
+	defer service.Clean()
+
+	requestPath := "/"
+	req, err := service.NewRequest("GET", requestPath)
+	if err != nil {
+		t.Fatalf("service.NewRequest: %v", err)
+	}
+
+	client := http.Client{Timeout: 10 * time.Second}
+	resp, err := client.Do(req)
+	if err != nil {
+		t.Fatalf("client.Do: %v", err)
+	}
+	defer resp.Body.Close()
+	fmt.Printf("client.Do: %s %s\n", req.Method, req.URL)
+
+	if got := resp.StatusCode; got != http.StatusOK {
+		t.Errorf("response status: got %d, want %d", got, http.StatusOK)
+	}
+}
+
+func GetLogEntries(service *cloudrunci.Service, runID string, projectID string, t *testing.T) {
+	ctx := context.Background()
+	client, err := logadmin.NewClient(ctx, projectID)
+	if err != nil {
+		t.Fatalf("Failed to create client: %v", err)
+	}
+	defer client.Close()
+
+	// Create service and timestamp filters
+	fiveMinAgo := time.Now().Add(-5 * time.Minute)
```

Comment:
If I recall correctly, there's not a strong guarantee of container scale-down happening in 5 minutes. This seems like a potential source of flakes. This test seems good to verify Cloud Run sends the SIGTERM && the SIGTERM is handled by the sample app.

Should we introduce a mechanism to trigger a SIGTERM so we can test the sample behavior more reliably (and quickly)?

Reply:
I have this setup for the service to be deleted then we start polling the logs, in order to force the scale-down. This has worked locally and on Kokoro. It is flaky on Kokoro, but that seems to be due to the logs being available.  Is there a time limit on how long we can wait for the logs?

Reply:
Yesterday there was an OMG for Cloud Logging, hence the logs seemings like they were not being flushes. As written this test max will take 5 minutes. It is passing now.